eval/src/vespa/eval/tensor/partial_update.cpp

```diff
@@ -0,0 +1,388 @@
+{
+    const ValueType &input_type = input.type();
+    const ValueType &modifier_type = modifier.type();
+    if (input_type.dimensions() != modifier_type.dimensions()) {
+        LOG(error, "when adding cells to a tensor, dimensions must be equal");
+        return Value::UP();
+    }
+    const auto input_cells = input.cells().typify<ICT>();
+    const auto modifier_cells = modifier.cells().typify<MCT>();
+    const size_t num_mapped_in_input = input_type.count_mapped_dimensions();
+    const size_t dsss = input_type.dense_subspace_size();
+    const size_t expected_subspaces = input.index().size() + modifier.index().size();
+    auto builder = factory.create_value_builder<ICT>(input_type, num_mapped_in_input, dsss, expected_subspaces);
+    Addresses addrs(num_mapped_in_input);
+    std::set<size_t> overwritten_subspaces;
+    auto modifier_view = modifier.index().create_view({});
+    auto lookup_view = input.index().create_view(addrs.lookup_view_dims);
+    modifier_view->lookup({});
+    size_t modifier_subspace_index;
+    while (modifier_view->next_result(addrs.next_result_refs, modifier_subspace_index)) {
+        size_t modifier_offset = dsss * modifier_subspace_index;
+        auto src = modifier_cells.begin() + modifier_offset;
+        auto dst = builder->add_subspace(addrs.addr).begin();
+        for (size_t i = 0; i < dsss; ++i) {
+            dst[i] = src[i];
+        }
+        lookup_view->lookup(addrs.lookup_refs);
+        size_t input_subspace_index;
+        if (lookup_view->next_result({}, input_subspace_index)) {
+            overwritten_subspaces.insert(input_subspace_index);
+        }
+    }
+    auto input_view = input.index().create_view({});
+    input_view->lookup({});
+    size_t input_subspace_index;
+    while (input_view->next_result(addrs.next_result_refs, input_subspace_index)) {
+        if (overwritten_subspaces.count(input_subspace_index) == 0) {
+            size_t input_offset = dsss * input_subspace_index;
+            auto src = input_cells.begin() + input_offset;
+            auto dst = builder->add_subspace(addrs.addr).begin();
+            for (size_t i = 0; i < dsss; ++i) {
+                dst[i] = src[i];
+            }
+        }
+    }
+    return builder->build(std::move(builder));
+}
+
+struct PerformAdd {
+    template<typename ICT, typename MCT>
+    static Value::UP invoke(const Value &input,
+                            const Value &modifier,
+                            const ValueBuilderFactory &factory)
+    {
+        return my_add_cells<ICT,MCT>(input, modifier, factory);
+    }
+};
+
+//-----------------------------------------------------------------------------
+
+template <typename ICT>
+Value::UP
+my_remove_cells(const Value &input, const Value &modifier, const ValueBuilderFactory &factory)
+{
+    const ValueType &input_type = input.type();
+    const ValueType &modifier_type = modifier.type();
+    if (input_type.mapped_dimensions() != modifier_type.mapped_dimensions()) {
```

Comment:
compare to modifier.dimensions() directly to ensure there are no dense ones as well.